src/meta/src/stream/mod.rs

```diff
@@ -20,9 +20,54 @@ mod stream_manager;
 
 #[cfg(test)]
 mod test_fragmenter;
-
 pub use meta::*;
 pub use scheduler::*;
 pub use source_manager::*;
 pub use stream_graph::*;
 pub use stream_manager::*;
+
+/// Set vnode mapping for stateful operators.
+#[macro_export]
+macro_rules! set_table_vnode_mappings {
+    () => {
+        fn set_table_vnode_mappings(
+            &self,
+            stream_node: &StreamNode,
+            fragment_id: FragmentId,
+        ) -> Result<()> {
+            match stream_node.get_node_body()? {
+                NodeBody::Materialize(node) => {
+                    let table_id = node.get_table_ref_id()?.get_table_id() as u32;
+                    self.hash_mapping_manager
+                        .set_fragment_state_table(fragment_id, table_id);
+                }
+                NodeBody::GlobalSimpleAgg(node) => {
+                    let table_ids = node.get_table_ids();
+                    for table_id in table_ids {
+                        self.hash_mapping_manager
+                            .set_fragment_state_table(fragment_id, *table_id);
+                    }
+                }
+                NodeBody::HashAgg(node) => {
+                    let table_ids = node.get_table_ids();
+                    for table_id in table_ids {
+                        self.hash_mapping_manager
+                            .set_fragment_state_table(fragment_id, *table_id);
+                    }
+                }
+                NodeBody::HashJoin(node) => {
+                    self.hash_mapping_manager
+                        .set_fragment_state_table(fragment_id, node.left_table_id);
+                    self.hash_mapping_manager
+                        .set_fragment_state_table(fragment_id, node.right_table_id);
+                }
```

Comment:
Are these all stateful operators? Do I omit any other? cc. @fuyufjh

Reply:
`TopN`?

Reply:
Since TopN is always singleton, it could be allowed not to have a vnode mapping. Compactor will omit the part of hash compaction according to the fact that vnode mapping is `None`, even if it **is** a stateful operator and should do hash compaction. Is this okay? cc. @hzxa21

Reply:
And maybe we could adopt same strategy on `GlobalSimpleAgg`?

Reply:
⬆️ LGTM. If so, `GlobalSimpleAgg` can also be omitted.

Reply:
> Since TopN is always singleton, it could be allowed not to have a vnode mapping. Compactor will omit the part of hash compaction according to the fact that vnode mapping is `None`, even if it **is** a stateful operator and **should** do hash compaction. Is this okay? cc. @hzxa21

+1. This makes sense to me. cc @soundOfDestiny what do you think?

Reply:
> > Since TopN is always singleton, it could be allowed not to have a vnode mapping. Compactor will omit the part of hash compaction according to the fact that vnode mapping is `None`, even if it **is** a stateful operator and **should** do hash compaction. Is this okay? cc. @hzxa21
> 
> +1. This makes sense to me. cc @soundOfDestiny what do you think?

currently when vnode mapping is None, its hash value will be assigned to a constant magic number set by zwang28